### PR TITLE
Tensorboard fix

### PR DESCRIPTION
**Pull request by Litchilitchy**

### CODE REVIEW

File: pyzoo/zoo/pipeline/api/keras/engine/topology.py [hunk 1]
```diff
@@ -74,6 +74,33 @@ def set_tensorboard(self, log_dir, app_name):
...
+                             self.value, tag, "Train")
+
+    def get_validation_summary(self, tag=None):
+        """
+        Get the scalar from model validation summary
+        Return 2-D array like object which could be converted
+        by nd.array()
+        # Arguments
+        tag: The string variable represents the scalar wanted
+        """
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/keras/engine/topology.py]: check tag?

Comment (Litchilitchy) [reply pyzoo/zoo/pipeline/api/keras/engine/topology.py]: @jason-dai fixed

File: pyzoo/zoo/pipeline/api/keras/engine/topology.py [hunk 2]
```diff
@@ -74,6 +74,40 @@ def set_tensorboard(self, log_dir, app_name):
...
+    def get_train_summary(self, tag=None):
+        """
+        Get the scalar from model train summary
+        Return 2-D array like object which could be converted
+        by nd.array()
+        # Arguments
+        tag: The string variable represents the scalar wanted
+        """
+        # exception handle
+        if tag != "Loss" and tag != "LearningRate" and tag != "Throughput":
```
Comment (hkvision) [inline pyzoo/zoo/pipeline/api/keras/engine/topology.py]: convert to lower case and check?

File: pyzoo/zoo/pipeline/api/keras/engine/topology.py [hunk 3]
```diff
@@ -74,6 +74,40 @@ def set_tensorboard(self, log_dir, app_name):
...
+                            + 'are supported in train summary')
+
+        return callBigDlFunc(self.bigdl_type, "zooGetScalarFromSummary",
+                             self.value, tag, "Train")
+
+    def get_validation_summary(self, tag=None):
+        """
+        Get the scalar from model validation summary
+        Return 2-D array like object which could be converted
+        by nd.array()
```
Comment (hkvision) [inline pyzoo/zoo/pipeline/api/keras/engine/topology.py]: nd.array or nd_array()?

File: pyzoo/zoo/pipeline/api/keras/engine/topology.py [hunk 4]
```diff
@@ -74,6 +74,40 @@ def set_tensorboard(self, log_dir, app_name):
...
+        # Arguments
+        tag: The string variable represents the scalar wanted
+        """
+        validation_set = set(('AUC', 'Accuracy', 'BinaryAccuracy', 'CategoricalAccuracy',
+                              'HitRatio', 'Loss', 'MAE', 'NDCG', 'SparseCategoricalAccuracy',
+                              'TFValidationMethod', 'Top1Accuracy',
+                              'Top5Accuracy', 'TreeNNAccuracy'))
+        if tag not in validation_set:
+            raise TypeError('Only subclasses of ValidationMethod are supported,'
```
Comment (hkvision) [inline pyzoo/zoo/pipeline/api/keras/engine/topology.py]: ValidationMethod is not exposed in Python and thus this error msg may be confusing to users?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 5]
```diff
@@ -184,12 +187,38 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
...
 
+  /**
+   * To get the scalar like "Loss", "LearningRate" from train summary
+   * Return is a Array of 3-tuples
+   *
+   * @param tag The string variable represents the parameter you want to return
+   *            supported tags are "LearningRate", "Loss", "Throughput"
+   */
+  def getTrainSummary(tag: String): Array[(Long, Float, Double)] = {
+    this.trainSummary.readScalar(tag)
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: will Scala do the tag check?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 6]
```diff
@@ -184,12 +187,38 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
...
+  }
+
+  /**
+   * To get the scalar like "Loss", "Top1Accuracy" from validation summary
+   * Return is a Array of 3-tuples
+   *
+   * @param tag The string variable represents the parameter you want to return
+   *            supported tags are "Loss", "Top1Accuracy", etc based on
+   *            validation metric. Use a method to_bigdl_metric to convert
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: this description seems to be not clear enough for users to understand?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala [hunk 7]
```diff
@@ -177,6 +177,26 @@ class PythonZooKeras[T: ClassTag](implicit ev: TensorNumeric[T]) extends PythonZ
...
 
+  def zooGetScalarFromSummary(
+      module: KerasNet[T],
+      tag: String,
+      target: String): JList[JList[Any]] = {
+
+    var scalarArray: Array[(Long, Float, Double)] = null
+    require(target == "Train" || target == "Validation",
+      "Invalid target, must be Train or Validation.")
+    if (target == "Train") {
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala]: change to the following?
```
val scalarArray = if ... else ...
```

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/models/TrainingSpec.scala [hunk 8]
```diff
@@ -224,6 +224,28 @@ class TrainingSpec extends FlatSpec with Matchers with BeforeAndAfter {
     res
   }
 
+  "tensorboard api" should "work" in {
+    val trainingData = generateData(Array(12, 12), 1, 100)
```
Comment (hkvision) [inline zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/models/TrainingSpec.scala]: need to add some unit test in Python as well.

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/models/TrainingSpec.scala [hunk 9]
```diff
@@ -224,6 +224,28 @@ class TrainingSpec extends FlatSpec with Matchers with BeforeAndAfter {
...
+
+    model.add(Dense[Float](8, activation = "relu", inputShape = Shape(12, 12)))
+    model.add(Flatten[Float]())
+    model.add(Dense[Float](2, activation = "softmax"))
+
+    model.compile(optimizer = "sgd", loss = "sparse_categorical_crossentropy",
+      metrics = List("accuracy"))
+    val api = new PythonZooKeras[Float]()
+
+    model.setTensorBoard("./", "testTensorBoard")
```
Comment (hkvision) [inline zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/models/TrainingSpec.scala]: better to remove the temporary test files?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 5 above)
Comment (zhichao-li) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: If you put the checking in scala, python can also take advantage of it.

File: pyzoo/zoo/pipeline/api/keras/engine/topology.py (hunk 3 above)
Comment (Litchilitchy) [reply pyzoo/zoo/pipeline/api/keras/engine/topology.py]: typo..

File: pyzoo/zoo/pipeline/api/keras/engine/topology.py (hunk 4 above)
Comment (Litchilitchy) [reply pyzoo/zoo/pipeline/api/keras/engine/topology.py]: so we output supported str (line 107)

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 5 above)
Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: I did not find check in scala. And currently most api do not check but few of them do. So I guess there is no general rule of whether python or scala to check?

Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: error messages from scala are not user friendly to python users?

Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Zhichao means if we put the check in Scala then we do not need to check in Python as they will eventually go to Scala.

Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: If we put `require tag==..` in scala with error msg, then it should be clear to Python users as well.

Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: according to this then we should support both scala and python check, thus all check should be done in scala because scala should reflect to python but it could not reverse. So why there exists some check in python?..

File: pyzoo/zoo/pipeline/api/keras/engine/topology.py (hunk 2 above)
Comment (Litchilitchy) [reply pyzoo/zoo/pipeline/api/keras/engine/topology.py]: or directly use lower case in python would be more user friendly?

Comment (hkvision) [reply pyzoo/zoo/pipeline/api/keras/engine/topology.py]: something like:
```
tag = tag.lower()
if tag != 'loss' ...
```??

File: pyzoo/test/zoo/pipeline/api/keras/test_simple_integration.py [hunk 10]
```diff
@@ -77,7 +77,11 @@ def test_training_with_tensorboard_checkpoint_gradientclipping(self):
         model.fit(X_train, y_train, batch_size=112, nb_epoch=2, validation_data=(X_test, y_test))
         model.set_gradient_clipping_by_l2_norm(0.2)
         model.fit(X_train, y_train, batch_size=112, nb_epoch=2, validation_data=(X_test, y_test))
-        model.evaluate(X_test, y_test, batch_size=112)
+        train_loss = model.get_train_summary("Loss")
+        val_loss = model.get_validation_summary("Loss")
+        np.array(train_loss)
+        np.array(val_loss)
+        eval = model.evaluate(X_test, y_test, batch_size=112)
```
Comment (Litchilitchy) [inline pyzoo/test/zoo/pipeline/api/keras/test_simple_integration.py]: python unit test directly added here

File: pyzoo/zoo/pipeline/api/keras/engine/topology.py (hunk 2 above)
Comment (Litchilitchy) [reply pyzoo/zoo/pipeline/api/keras/engine/topology.py]: so case not sensitive? support sth like "LoSs"... may we directly force all params to be lower case? since other params in python seems all to be lower case.

Comment (Litchilitchy) [reply pyzoo/zoo/pipeline/api/keras/engine/topology.py]: and if so, the validation method becomes sth like `sparse_categorical_entropy` using lower case and  `_`
